### PR TITLE
SG-43456 [Phases 1-2] Enable FlowAM integration

### PR DESCRIPTION
**Pull request by carlos-villavicencio-adsk**
Base branch for tk-multi-loader2 v2.0.0

## Approved Contents

- MEDM Abstractions
- #138 
- #139 
- #140 
- #142 
- #143
- #144 
- #145
- #147 
- #148

ℹ️ This should not affect or break the loader functionality.

## Dependencies

- tk-core:  https://github.com/shotgunsoftware/tk-core/pull/1108
- tk-maya:  https://github.com/shotgunsoftware/tk-maya/pull/130
- tk-houdini:  https://github.com/shotgunsoftware/tk-houdini/pull/103
- tk-nuke:  https://github.com/shotgunsoftware/tk-nuke/pull/134
- tk-desktop:  https://github.com/shotgunsoftware/tk-desktop/pull/224

## Config Requirements

- core/core_api.yml
    - Use a descriptor to `master`

## TODO

- [x] Remove the `tk-framework-flowam` entry on `info.yml` that is causing CI to fail #145 
- [x] Add flowam config instructions
- [x] Tested on Maya
- [x] Tested on Desktop
- [x] Remove `tk_core_ref` from `azure-pipelines.yml` after https://github.com/shotgunsoftware/tk-core/pull/1108 is merged

### CODE REVIEW

File: python/tk_multi_loader/flowam/publishhistory_model.py [hunk 1]
```diff
@@ -0,0 +1,507 @@
...
+        qt_item.get_sg_data = get_sg_data
+
+        # Prepend so drafts always appear above all published versions.
+        self.insertRow(0, qt_item)
+        self._app.log_debug(
+            f"MEDM History: Added {draft_type} draft '{draft_info.name}' "
+            f"(draft_id={draft_info.draft_id})"
+        )
+
+    def _resolve_publish_type(self, medm_type_id_str: str) -> tuple:
```
Comment (chenm1adsk) [inline python/tk_multi_loader/flowam/publishhistory_model.py:402]: This function exists in latestpublish_model.py as well. Should we move it to utils.py to avoid duplicate?

File: python/tk_multi_loader/medm/publishhistory_model.py [hunk 2]
```diff
@@ -0,0 +1,507 @@
...
+            "id": None,
+            "type": "PublishedFile",
+            "code": asset_revision.name,
+            "name": asset_revision.name,
+            "version_number": version.version_number,
+            "description": description,
+            "created_at": version.created_at,
+            "created_by": {
+                "type": "HumanUser",
+                "id": 1,
```
Comment (chenm1adsk) [inline python/tk_multi_loader/medm/publishhistory_model.py]: Sorry could you please remind me why this id for create_by is 1?

File: python/tk_multi_loader/flowam/thumbnail_service.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+            self._timer.timeout.connect(self._process_pending)
+            self._timer.start(100)
+
+    def _resolve_and_fetch(
+        self, qt_item: QtGui.QStandardItem, revision_id: str, callback: Callable
+    ) -> None:
+        """Background-thread worker: resolve URL then download bytes."""
+        # 1. Resolve the thumbnail URL (hits the API at most once per revision).
+        url = self._url_cache.get(revision_id)
+        if url is None and revision_id not in self._url_cache:
```
Comment (chenm1adsk) [inline python/tk_multi_loader/flowam/thumbnail_service.py:150]: Is this just equivalent to `if revision_id not in self._url_cache`?

**Review by chenm1adsk — COMMENTED** [review]
Left some small comments and other looks good to me.

File: python/tk_multi_loader/medm/publishhistory_model.py (hunk 2 above)
Comment (carlos-villavicencio-adsk) [reply python/tk_multi_loader/medm/publishhistory_model.py]: It's a mocked value. But it's not ok to be that magical. I'll try to set it to `None` which is better.

File: python/tk_multi_loader/flowam/thumbnail_service.py (hunk 3 above)
Comment (carlos-villavicencio-adsk) [reply python/tk_multi_loader/flowam/thumbnail_service.py:150]: I believe we're checking the `url` value as well  because it can actually contain a cached value.

Comment (chenm1adsk) [reply python/tk_multi_loader/flowam/thumbnail_service.py:150]: Isn't that url = self._url_cache.get(revision_id)? so if revision_id not in self._url_cache url is None?

**Review by chenm1adsk — APPROVED** [review]

File: hooks/tk-desktop_actions.py [hunk 4]
```diff
@@ -76,7 +76,9 @@ def generate_actions(
 
         action_instances = []
 
-        if app.context.flow_project_id and sgtk.platform.current_engine().flow_host:
+        if hasattr(app.context, "flow_project_id") and hasattr(
+            sgtk.platform.current_engine(), "flow_host"
+        ):
```
Comment (yungsiow) [inline hooks/tk-desktop_actions.py]: You also need to test that these values are not None.  It is not enough that the properties exist.

- properties exist => tk-core version is new enough
- values are not None => we're in a flow configured project

Comment (yungsiow) [reply hooks/tk-desktop_actions.py]: Might be worth wrapping this in a small helper function.

Comment (carlos-villavicencio-adsk) [reply hooks/tk-desktop_actions.py]: Fixed on 110faebcfc0cdae7169235f032da3a682136a9b9

**Review by yungsiow — APPROVED** [review]
Looks good!
